### PR TITLE
Disallow changes to immutable executables or files in them. Fixes #1796.

### LINKED ISSUE #1796: Deleting a file from an executable doesn't update the immutable executable hash
Steps to reproduce:

- Write down the hash for an immutable executable
- Delete a file from an executable
- Observe that no new immutable executable is created nor are the hashes updated

This will later break judging when the judgehost compares the hash.

### PR DESCRIPTION
**Pull request by nickygerritsen**
I did the thing Jaap suggested and removed all writable methods on the immutable executable entity.

However, we removed executable files from the DB directly, so I also forbid that.

### CODE REVIEW

File: webapp/src/Controller/Jury/ExecutableController.php [hunk 1]
```diff
@@ -268,7 +268,26 @@ public function deleteSingleAction(Request $request, string $execId, int $rank):
 
             return $this->render('jury/delete.html.twig', $data);
         } else {
-            $this->em->remove($file);
+            // Create a copy of all files except $file
```
Comment (meisterT) [inline webapp/src/Controller/Jury/ExecutableController.php:271]: The `$file` in this comment is the `$file from above` and then in the loop below is a different `$file`. While this is valid code, it is confusing.

I suggest renaming `$file` to `$fileToDelete`, and `$rank` to `$rankToDelete`.

File: webapp/src/Entity/ExecutableFile.php [hunk 2]
```diff
@@ -138,4 +139,12 @@ public function isExecutable(): bool
     {
         return $this->isExecutable;
     }
+
+    /**
+     * @ORM\PreRemove()
+     */
+    public function disallowDelete(): void
+    {
+        throw new \RuntimeException('An executable file cannot be deleted');
```
Comment (meisterT) [inline webapp/src/Entity/ExecutableFile.php:148]: If you delete an immutable executable, can you still delete the file? Should you be able to even?

Comment (nickygerritsen) [reply webapp/src/Entity/ExecutableFile.php:148]: You can't delete an immutable executable either.

File: webapp/src/Controller/Jury/ExecutableController.php [hunk 3]
```diff
@@ -268,7 +268,26 @@ public function deleteSingleAction(Request $request, string $execId, int $rank):
...
+            /** @var ExecutableFile $file */
+            foreach ($executable->getImmutableExecutable()->getFiles() as $file) {
+                if ($file->getRank() == $rankToDelete) {
+                    continue;
+                }
+
+                $executableFile = new ExecutableFile();
+                $executableFile
+                    ->setRank($file->getRank())
+                    ->setIsExecutable($file->isExecutable())
```
Comment (vmcj) [inline webapp/src/Controller/Jury/ExecutableController.php:282]: When is an executableFile not Executable?

File: webapp/src/Entity/ImmutableExecutable.php [hunk 4]
```diff
@@ -33,7 +37,7 @@ class ImmutableExecutable
      * @ORM\JoinColumn(name="userid", referencedColumnName="userid", onDelete="SET NULL")
```
Comment (vmcj) [inline webapp/src/Entity/ImmutableExecutable.php:37]: The `SET NULL` is still needed because we disconnect the user from the Executable?

Comment (nickygerritsen) [reply webapp/src/Entity/ImmutableExecutable.php:37]: I don’t know. I don’t know when we filled it

**Review by meisterT — APPROVED** [review]

**Pull request merged**